sklearn/utils/estimator_checks.py

```diff
@@ -2030,7 +2033,11 @@ def param_filter(p):
             if isinstance(param_value, np.ndarray):
                 assert_array_equal(param_value, init_param.default)
             else:
-                assert_equal(param_value, init_param.default, init_param.name)
+                # Allows to set default parameters to np.nan
+                if (param_value is not np.nan or
+                        init_param.default is not np.nan):
+                    assert_equal(param_value, init_param.default,
+                                 init_param.name)
```

Comment:
I think this if block should be replaced by:

```python
if param_value is np.nan:
    assert param_value is init_param.default, init_param.name
else:
    assert param_value == init_param.default, init_param.name
```

pytest should be able to yield informative assertion error message displaying both values and the name of the parameter with this simple construct.

Reply:
Actually we need:

```python
def isscalarnan(x):
    return isinstance(x, numbers.Real) and np.isnan(x)

if isscalarnan(param_value):
    assert param_value is init_param.default, init_param.name
else:
    assert param_value == init_param.default, init_param.name
```

Reply:
what about if param_value != param_value ?
this always works

Reply:
I think I like the `isscalarnan` variant better as I find it more explicit.

Reply:
The above block with the `assert_equal` should be removed as it's now redundant with the:

```python
assert param_value == init_param.default, init_param.name
```

that you added below in the else clause.